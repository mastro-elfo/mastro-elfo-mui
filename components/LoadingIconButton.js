"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = LoadingIconButton;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _notistack = require("notistack");

var _core = require("@material-ui/core");

var _ = require("./");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function LoadingIconButton(_ref) {
  var children = _ref.children,
      _ref$callback = _ref.callback,
      callback = _ref$callback === void 0 ? function () {
    return Promise.reject(new Error("No callback provided"));
  } : _ref$callback,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      _ref$AbsoluteCircular = _ref.AbsoluteCircularProgressProps,
      AbsoluteCircularProgressProps = _ref$AbsoluteCircular === void 0 ? {} : _ref$AbsoluteCircular,
      others = _objectWithoutProperties(_ref, ["children", "callback", "disabled", "AbsoluteCircularProgressProps"]);

  var _useSnackbar = (0, _notistack.useSnackbar)(),
      enqueueSnackbar = _useSnackbar.enqueueSnackbar;

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      loading = _useState2[0],
      setLoading = _useState2[1];

  var handleClick = function handleClick() {
    setLoading(true);
    callback()["catch"](function (err) {
      console.error(err);
      enqueueSnackbar(err.message, {
        variant: "error"
      });
    }).then(function () {
      setLoading(false);
    });
  };

  return /*#__PURE__*/_react["default"].createElement(_core.IconButton, _extends({
    disabled: loading || disabled,
    onClick: handleClick
  }, others), children, /*#__PURE__*/_react["default"].createElement(_.Loading, {
    loading: loading
  }, /*#__PURE__*/_react["default"].createElement(_.AbsoluteCircularProgress, _extends({
    color: "secondary"
  }, AbsoluteCircularProgressProps))));
}

LoadingIconButton.propTypes = {
  children: _propTypes["default"].node,
  callback: _propTypes["default"].func,
  disabled: _propTypes["default"].bool,
  AbsoluteCircularProgressProps: _propTypes["default"].object
};