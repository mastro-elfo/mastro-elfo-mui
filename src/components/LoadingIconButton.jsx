import React, { useState } from "react";
import PropTypes from "prop-types";

import { useSnackbar } from "notistack";

import { IconButton } from "@material-ui/core";

import AbsoluteCircularProgress from "./AbsoluteCircularProgress";
import Loading from "./Loading";

/**
 * `IconButton` with `Loading` component
 * @param node      children
 * @param function  callback Deprecated
 * @param Boolean   disabled=false  [description]
 * @param function  onClick=()=>{} Callback on click event
 * @param Object    AbsoluteCircularProgressProps={} Properties for `AbsoluteCircularProgress` component
 * @param Object    n sLoadingProps={}  Properties for `Loading` component
 * @constructor
 */
export default function LoadingIconButton({
  children,
  callback,
  disabled = false,
  onClick,
  AbsoluteCircularProgressProps = {},
  LoadingProps = {},
  ...others
}) {
  const { enqueueSnackbar } = useSnackbar();
  const [loading, setLoading] = useState(false);

  if (callback !== undefined) {
    console.warn(
      "`callback` is deprecated since v1.24.0 and will be removed in v2.0.0. Use `onClick` instead"
    );
  }

  const cb =
    onClick !== undefined
      ? onClick
      : callback !== undefined
      ? callback
      : () => Promise.reject(new Error("No callback provided"));

  const handleClick = () => {
    setLoading(true);
    Promise.resolve()
      .then(() => cb())
      .catch(err => {
        console.error(err);
        enqueueSnackbar(err.message, { variant: "error" });
      })
      .then(() => {
        setLoading(false);
      });

    // setLoading(true);
    // callback()
    //   .catch(err => {
    //     console.error(err);
    //     enqueueSnackbar(err.message, { variant: "error" });
    //   })
    //   .then(() => {
    //     setLoading(false);
    //   });
  };

  return (
    <IconButton
      disabled={loading || disabled}
      onClick={handleClick}
      {...others}
    >
      {children}
      <Loading loading={loading} {...LoadingProps}>
        <AbsoluteCircularProgress
          color="secondary"
          {...AbsoluteCircularProgressProps}
        />
      </Loading>
    </IconButton>
  );
}

LoadingIconButton.propTypes = {
  children: PropTypes.node,
  callback: PropTypes.func,
  disabled: PropTypes.bool,
  AbsoluteCircularProgressProps: PropTypes.object
};
