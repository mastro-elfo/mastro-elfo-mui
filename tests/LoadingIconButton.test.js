import React from "react";
import renderer from "react-test-renderer";

import { SnackbarProvider } from "notistack";
import LoadingIconButton from "../src/components/LoadingIconButton";

test("", () => {
  const component = renderer.create(
    <SnackbarProvider>
      <LoadingIconButton></LoadingIconButton>
    </SnackbarProvider>
  );
  let tree = component.toJSON();
  // console.log(tree);
  expect(tree).toMatchSnapshot();
});
