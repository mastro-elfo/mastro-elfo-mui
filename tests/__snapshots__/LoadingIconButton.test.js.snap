// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<button
  className="MuiButtonBase-root MuiIconButton-root"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiIconButton-label"
  />
</button>
`;
